internal/util/secretskms.go

```diff
@@ -0,0 +1,224 @@
+func (kms SecretsKMS) StoreDEK(key, value string) error {
+	return nil
+}
+
+// RemoveDEK is doing nothing as no new passphrases are saved with
+// SecretsKMS.
+func (kms SecretsKMS) RemoveDEK(key string) error {
+	return nil
+}
+
+// SecretsMetadataKMS is a KMS based on the SecretsKMS, but stores the
+// Data-Encryption-Key (DEK) in the metadata of the volume.
+type SecretsMetadataKMS struct {
+	SecretsKMS
+
+	encryptionKMSID string
+}
+
+// initSecretsMetadataKMS initializes a SecretsMetadataKMS that wraps a
+// SecretsKMS, so that the passphrase from the StorageClass secrets can be used
+// for encrypting/decrypting DEKs that are stored in a detached DEKStore.
+func initSecretsMetadataKMS(encryptionKMSID string, secrets map[string]string) (EncryptionKMS, error) {
+	eKMS, err := initSecretsKMS(secrets)
+	if err != nil {
+		return nil, err
+	}
+
+	sKMS, ok := eKMS.(SecretsKMS)
+	if !ok {
+		return nil, fmt.Errorf("failed to convert %T to SecretsKMS", eKMS)
+	}
+
+	smKMS := SecretsMetadataKMS{}
+	smKMS.SecretsKMS = sKMS
+	smKMS.encryptionKMSID = encryptionKMSID
+
+	return smKMS, nil
+}
+
+// GetID is returning ID representing the SecretsMetadataKMS.
+func (kms SecretsMetadataKMS) GetID() string {
+	return kms.encryptionKMSID
+}
+
+// Destroy frees all used resources.
+func (kms SecretsMetadataKMS) Destroy() {
+	kms.SecretsKMS.Destroy()
+}
+
+func (kms SecretsMetadataKMS) requiresDEKStore() DEKStoreType {
+	return DEKStoreMetadata
+}
+
+// encryptedMetedataDEK contains the encrypted DEK and the Nonce that was used
+// during encryption. This structure is stored (in JSON format) in the DEKStore
+// that is linked to this KMS provider.
+type encryptedMetedataDEK struct {
+	// DEK is the encrypted data-encryption-key for the volume.
+	DEK []byte `json:"dek"`
+	// Nounce is a random byte slice to guarantee the uniqueness of the
+	// encrypted DEK.
+	Nonce []byte `json:"nonce"`
+}
+
+// EncryptDEK encrypts the plainDEK with a key derived from the passphrase from
+// the SecretsKMS and the volumeID.
+// The resulting encryptedDEK contains a JSON with the encrypted DEK and the
+// nonce that was used for encrypting.
+func (kms SecretsMetadataKMS) EncryptDEK(volumeID, plainDEK string) (string, error) {
+	// use the passphrase from the SecretsKMS
+	passphrase, err := kms.SecretsKMS.FetchDEK(volumeID)
+	if err != nil {
+		return "", fmt.Errorf("failed to get passphrase: %w", err)
+	}
+
+	aead, err := generateCipher(passphrase, volumeID)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate cipher: %w", err)
+	}
+
+	emd := encryptedMetedataDEK{
+		Nonce: generateNonce(aead.NonceSize()),
+	}
+	emd.DEK = aead.Seal(nil, emd.Nonce, []byte(plainDEK), nil)
+
+	emdData, err := json.Marshal(&emd)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert "+
+			"encryptedMetedataDEK to JSON: %w", err)
+	}
+
+	return string(emdData), nil
+}
+
+// DecryptDEK takes the JSON formatted `encryptedMetedataDEK` contents, and the
```

Comment:
and `it` ( instead of `the`) fetches ... ?

Reply:
yes, will fix